collections/_posts/2024-08-26-SAM2.md

```diff
@@ -0,0 +1,170 @@
+
+Both aspects will be presented in the next sections.
+
+&nbsp;
+
+# Segment Anything Model 2
+
+<div style="text-align:center">
+<img src="/collections/images/SAM2/sam2_overview.jpg" width=1500></div>
+<p style="text-align: center;font-style:italic">Figure 2. Overview of SAM 2.</p>
+
+SAM 2 is broadly similar to SAM with two major changes : a different, more recent and powerful image encoder, and a *memory mechanism* for ensuring consistency between segmented masks across frames. As a reminder, SAM is composed of an ViT image encoder, a prompt encoder and a lightweight decoder for producing the output masks. A consequence of such an asymetric model is that encoding an input image is rather long, but adjusting the prompts is then very fast. This property is also found in SAM 2. Figure 2 details all the components of the model.
+
+&nbsp;
+
+## Image encoder
+
+They authors used a **Hiera**[^1], a recent hierarchical vision transformer using only standard ViT blocks and global attention (except for the first two stages) and designed to be more efficient than windowed SwinViT or MViT. The image encoder is pretrained with Masked Auto-Encoder procedure. The final frame embeddings are obtained by fusing the features from stages 3 and 4 of the Hira image encoder. Contrary to SAM, **intermediate stage features are also kept** and later added to the upsampling layers in the mask decoder to help producing high-resolution segmentation details.
+
+The image encoder is run only once over all the images to produce *unconditioned image tokens* representing each frame.
+
+&nbsp;
+
+## Memory attention
+
+The role of memory attention is to **condition frame features on the past feature frames and predictions** as well as on any new prompts. It is composed of several transformer blocks in which there is self-attention between the image embeddings of the current frame, and cross-attention to tokens of the objects in the memory bank, that is prompted or unprompted frames and objets pointers (see relevant section for further details).
+
+&nbsp;
+
+## Prompt encoder
+
+Prompt encoders are similar to SAM's : sparse prompts, *i.e.* points and bounding boxes, are represented by positional encodings summed with learnt embeddings for each prompt type and dense prompts, *i.e.* masks, are embedded using (downsampling) convolutions and summed with the conditioned frame embeddings.
+
+&nbsp;
+
+## Mask decoder
+
+The decoder architecture is again quite similar to that of SAM. It takes as input the conditioned frame embeddings (with potential mask encoded embeddings) on one hand, and sparse prompts tokens, concatenated with tokens used for the output, on the other hand. The decoder stacks transformer blocks with self-attention between prompt+ouput tokens and alternated "image to token" and "token to image" cross-attention.
```

Comment:
The last sentence of this paragraph (about the stack of transformer blocks in the decoder) is confusing to me. I feel like there's too much info in it, so maybe try splitting into a few simpler descriptions?

Reply:
see latest commit